core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/LDAPConnectionContext.java

```diff
@@ -203,10 +209,87 @@ public LDAPConnectionContext(RealmConfiguration realmConfig) throws UserStoreExc
         // Set StartTLS option if provided in the configuration. Otherwise normal connection.
         startTLSEnabled = Boolean.parseBoolean(realmConfig.getUserStoreProperty(
                 UserStoreConfigConstants.STARTTLS_ENABLED));
+
+        // Set waiting time to re-establish LDAP connection after couple of failure attempts if specified otherwise
+        // set default time.
+        String retryWaitingTime = realmConfig.getUserStoreProperty(UserStoreConfigConstants.CONNECTION_RETRY_DELAY);
+        if (StringUtils.isNotEmpty(retryWaitingTime)) {
+            thresholdTimeoutInMilliseconds = getThresholdTimeoutInMilliseconds(retryWaitingTime);
+        } else {
+            thresholdTimeoutInMilliseconds = 120000;
+        }
+        // By-default set to close state.
+        ldapConnectionCircuitBreakerState = CIRCUIT_STATE_CLOSE;
+        thresholdStartTime = 0;
     }
 
     public DirContext getContext() throws UserStoreException {
 
+        DirContext context = null;
+
+        // Implemented basic circuit breaker logic to reduce the resource consumption.
+        switch (ldapConnectionCircuitBreakerState) {
+        case CIRCUIT_STATE_OPEN:
+            long circuitOpenDuration = System.currentTimeMillis() - thresholdStartTime;
+            if (circuitOpenDuration >= thresholdTimeoutInMilliseconds) {
+                try {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Trying to obtain LDAP connection, connection URL: " + environment
+                                .get(Context.PROVIDER_URL) + " when circuit breaker state: "
+                                + ldapConnectionCircuitBreakerState + " and circuit breaker open duration: "
+                                + circuitOpenDuration + "ms.");
+                    }
+                    context = getDirContext();
+                    ldapConnectionCircuitBreakerState = CIRCUIT_STATE_CLOSE;
+                    thresholdStartTime = 0;
+                    break;
+                } catch (UserStoreException e) {
+                    log.error("Error occurred while obtaining LDAP connection. Connection URL: " + environment
+                            .get(Context.PROVIDER_URL), e);
+                    thresholdStartTime = System.currentTimeMillis();
+                    if (log.isDebugEnabled()) {
+                        log.debug("LDAP connection circuit breaker state set to: " + ldapConnectionCircuitBreakerState);
+                    }
+                    throw new UserStoreException("Error occurred while obtaining LDAP connection.", e);
+                }
+            } else {
```

Comment:
Shall we add some debug logs to else part as well.?

Reply:
following exception throws detailed error message hence avoided adding debug logs.